### PR TITLE
Bugfix: prevent daemon controller to adopt controller revisions of ot…

### LINKED ISSUE #101249: DaemonSet controller can adopt ControllerRevisions in other namespaces
#### What happened:
DaemonSet A can adopt orphan ControllerRevision resources of Daemonset B in other namespaces.

It can happen if DaemonSet A and B has the same "Spec.Selector" field.

#### What you expected to happen:
DaemonSet Controller should only adopt ControllerRevision in it's own namespace.

#### How to reproduce it (as minimally and precisely as possible):
1. create DaemonSet testns1/test-ds-in-testns1

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: test-ds-in-testns1
  namespace: testns1
spec:
  selector:
    matchLabels:
      name: fluentd-elasticsearch
  template:
    metadata:
      labels:
        name: fluentd-elasticsearch
    spec:
      containers:
      - name: fluentd-elasticsearch
        image: quay.io/fluentd_elasticsearch/fluentd:v2.5.2
        resources:
          limits:
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
```

2.  create DaemonSet testns2/test-ds-in-testns2, having same spec.selector with testns1/test-ds-in-testns1

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: test-ds-in-testns2
  namespace: testns2
spec:
  selector:
    matchLabels:
      name: fluentd-elasticsearch
  template:
    metadata:
      labels:
        name: fluentd-elasticsearch
    spec:
      containers:
      - name: fluentd-elasticsearch
        image: quay.io/fluentd_elasticsearch/fluentd:v2.5.2
        resources:
          limits:
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
```

3.  kubectl delete ds test-ds-in-testns1 -n testns1 --cascade=false
4.  kubectl label ds  test-ds-in-testns2 -n testns2 a=1 // to ensure this ds enqueue
5.  kubectl get controllerrevision -n testns1 -o wide
```
NAME                            CONTROLLER                          REVISION   AGE
test-ds-in-testns1-796898b788   daemonset.apps/test-ds-in-testns2   1          2m24s
```

#### Anything else we need to know?:
If this situation meets the case which is fixed by this [PR](https://github.com/kubernetes/kubernetes/pull/92743),  
unexpected deletions of A's children(Pods, ControllerRevisions and so on)  may happen, which can be much serious.

#### Environment:
- Kubernetes version (use `kubectl version`): v1.16.3-594
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by evertrain**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Solve DaemonSet Controller bug

#### Which issue(s) this PR fixes:
Fixes #101249

#### Special notes for your reviewer:
If the bug meets the case which is fixed by this [PR](https://github.com/kubernetes/kubernetes/pull/92743),
unexpected deletions of A's children(Pods, ControllerRevisions and so on) may happen, which can be much serious.

```release-note
NONE
```

### CODE REVIEW

**Comment by evertrain** [comment]
I signed it

File: pkg/controller/daemon/update.go [hunk 1]
```diff
@@ -399,7 +399,7 @@ func (dsc *DaemonSetsController) controlledHistories(ds *apps.DaemonSet) ([]*app
 
 	// List all histories to include those that don't match the selector anymore
 	// but have a ControllerRef pointing to the controller.
-	histories, err := dsc.historyLister.List(labels.Everything())
+	histories, err := dsc.historyLister.ControllerRevisions(ds.Namespace).List(labels.Everything())
```
Comment (krmayankk) [inline pkg/controller/daemon/update.go:402]: this looks like in the right direction(based on the description in the bug)

File: pkg/controller/daemon/update_test.go [hunk 2]
```diff
@@ -634,3 +635,73 @@ func TestGetUnavailableNumbers(t *testing.T) {
...
+					t.Fatalf("error creating DaemonSets controller: %v", err)
+				}
+				manager.dsStore.Add(ds1)
+				return manager
+			}(),
+			historyCRAll:              []*apps.ControllerRevision{crOfDs1, orphanCrInSameNsWithDs1},
+			expectControllerRevisions: []*apps.ControllerRevision{crOfDs1, orphanCrInSameNsWithDs1},
+		},
+		{
+			name: "Skip adopt the controller revision in the other namespaces",
```
Comment (krmayankk) [inline pkg/controller/daemon/update_test.go:665]: nit: Skip adopting the controller revision in namespace other than the one in which DS lives

File: pkg/controller/daemon/update_test.go [hunk 3]
```diff
@@ -634,3 +635,73 @@ func TestGetUnavailableNumbers(t *testing.T) {
...
+		}
+		if len(crList) != len(c.expectControllerRevisions) {
+			t.Errorf("Test case: %s, expect controllerrevision count %d but got:%d",
+				c.name, len(c.expectControllerRevisions), len(crList))
+		} else {
+			// check controller revisions match
+			for _, cr := range crList {
+				found := false
+				for _, expectCr := range c.expectControllerRevisions {
+					if reflect.DeepEqual(cr, expectCr) {
```
Comment (krmayankk) [inline pkg/controller/daemon/update_test.go:694]: will this verify same namespace as well

**Review by krmayankk — APPROVED** [review]
this looks good to me, but i will appreciate other people looking at it as well
@kubernetes/sig-apps-pr-reviews

**Comment by krmayankk** [comment]
not adding the lgtm label, waiting for more reviews from the team

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
Thanks for the fix!

**Comment by evertrain** [comment]
> not adding the lgtm label, waiting for more reviews from the team

@krmayankk  please merge this pr, thanks

**Comment by evertrain** [comment]
> Thanks for the fix!

@janetkuo need lgtm ;)

**Comment by krmayankk** [comment]
@evertrain  can you answer the open comments ?

File: pkg/controller/daemon/update.go (hunk 1 above)
Comment (evertrain) [reply pkg/controller/daemon/update.go:402]: yes，controllerrevision is a namespaced resource

File: pkg/controller/daemon/update_test.go (hunk 2 above)
Comment (evertrain) [reply pkg/controller/daemon/update_test.go]: fixed

File: pkg/controller/daemon/update_test.go (hunk 3 above)
Comment (evertrain) [reply pkg/controller/daemon/update_test.go:694]: yes, this will check them "deeply equal", including namespace, name and so on.

**Comment by evertrain** [comment]
> @evertrain can you answer the open comments ?

replied, thanks

File: pkg/controller/daemon/update.go (hunk 1 above)
Comment (liggitt) [reply pkg/controller/daemon/update.go:402]: this looks correct. can we also add safeguards into the owner/ownerref controller helpers (e.g. ClaimControllerRevisions, NewControllerRevisionControllerRefManager) to make adding cross-namespace owner references impossible?

Comment (evertrain) [reply pkg/controller/daemon/update.go:409]: > this looks correct. can we also add safeguards into the owner/ownerref controller helpers (e.g. ClaimControllerRevisions, NewControllerRevisionControllerRefManager) to make adding cross-namespace owner references impossible?

good suggestion. add namespace check in BaseControllerRefManager.ClaimObject, please review again

File: pkg/controller/controller_ref_manager.go [hunk 4]
```diff
@@ -107,6 +107,12 @@ func (m *BaseControllerRefManager) ClaimObject(ctx context.Context, obj metav1.O
 		// Ignore if the object is being deleted
 		return false, nil
 	}
+
+	if m.Controller.GetNamespace() != obj.GetNamespace() {
```
Comment (liggitt) [inline pkg/controller/controller_ref_manager.go:111]: only apply this check if the controller is actually namespaced:
```suggestion
	if len(m.Controller.GetNamespace()) > 0 && m.Controller.GetNamespace() != obj.GetNamespace() {
```

File: pkg/controller/controller_ref_manager.go [hunk 5]
```diff
@@ -107,6 +107,12 @@ func (m *BaseControllerRefManager) ClaimObject(ctx context.Context, obj metav1.O
 		// Ignore if the object is being deleted
 		return false, nil
 	}
+
+	if len(m.Controller.GetNamespace()) > 0 && m.Controller.GetNamespace() != obj.GetNamespace() {
```
Comment (liggitt) [inline pkg/controller/controller_ref_manager.go:111]: is there a unit test exercising adoption of a namespaced pod by a cluster-scoped controller?

**Pull request merged**